bridges/snowbridge/pallets/inbound-queue-v2/src/lib.rs

```diff
@@ -244,63 +197,46 @@ pub mod pallet {
 			// Verify that the message was submitted from the known Gateway contract
 			ensure!(T::GatewayAddress::get() == envelope.gateway, Error::<T>::InvalidGateway);
 
-			// Retrieve the registered channel for this message
-			let channel =
-				T::ChannelLookup::lookup(envelope.channel_id).ok_or(Error::<T>::InvalidChannel)?;
-
-			// Verify message nonce
-			<Nonce<T>>::try_mutate(envelope.channel_id, |nonce| -> DispatchResult {
-				if *nonce == u64::MAX {
-					return Err(Error::<T>::MaxNonceReached.into())
-				}
-				if envelope.nonce != nonce.saturating_add(1) {
-					Err(Error::<T>::InvalidNonce.into())
-				} else {
-					*nonce = nonce.saturating_add(1);
-					Ok(())
-				}
-			})?;
-
-			// Reward relayer from the sovereign account of the destination parachain, only if funds
-			// are available
-			let sovereign_account = sibling_sovereign_account::<T>(channel.para_id);
-			let delivery_cost = Self::calculate_delivery_cost(message.encode().len() as u32);
-			let amount = T::Token::reducible_balance(
-				&sovereign_account,
-				Preservation::Preserve,
-				Fortitude::Polite,
-			)
-			.min(delivery_cost);
-			if !amount.is_zero() {
-				T::Token::transfer(&sovereign_account, &who, amount, Preservation::Preserve)?;
-			}
+			// Verify the message has not been processed
+			ensure!(!<Nonce<T>>::contains_key(envelope.nonce), Error::<T>::InvalidNonce);
 
-			// Decode payload into `VersionedMessage`
-			let message = VersionedMessage::decode_all(&mut envelope.payload.as_ref())
+			// Decode payload into `MessageV2`
+			let message = MessageV2::decode_all(&mut envelope.payload.as_ref())
 				.map_err(|_| Error::<T>::InvalidPayload)?;
 
-			// Decode message into XCM
-			let (xcm, fee) = Self::do_convert(envelope.message_id, message.clone())?;
+			// Decode xcm
+			let versioned_xcm = VersionedXcm::<()>::decode_with_depth_limit(
+				MAX_XCM_DECODE_DEPTH,
+				&mut message.xcm.as_ref(),
+			)
+			.map_err(|_| Error::<T>::InvalidPayload)?;
+			let xcm: Xcm<()> = versioned_xcm.try_into().map_err(|_| <Error<T>>::InvalidPayload)?;
 
 			log::info!(
 				target: LOG_TARGET,
-				"💫 xcm decoded as {:?} with fee {:?}",
+				"💫 xcm decoded as {:?}",
 				xcm,
-				fee
 			);
 
-			// Burning fees for teleport
-			Self::burn_fees(channel.para_id, fee)?;
+			// Set nonce flag to true
+			<Nonce<T>>::try_mutate(envelope.nonce, |done| -> DispatchResult {
+				*done = true;
+				Ok(())
+			})?;
+
+			// Todo: Deposit fee(in Ether) to RewardLeger which should cover all of:
+			// T::RewardLeger::deposit(who, envelope.fee.into())?;
+			// a. The submit extrinsic cost on BH
+			// b. The delivery cost to AH
```

Comment:
The xcm can be built here without the need for XCM v5 and we can re-do later when XCM v5 is ready.

You will need to parse the Assets in the ticket and build the following XCM to send to asset hub:
```
ReceiveTeleportedAsset(asset_hub_fees)
BuyExecution(fees: asset_hub_fees)
DescendOrigin(PalletInstance(inbound_queue_pallet_index))
UniversalOrigin(GlobalConsensus(network))
ReserveAssetDeposited(erc20_assets)    // from assets that do not have tokenId
WithdrawAsset(pna_assets)                      // from assets that do have tokenId
DescendOrigin(msg_sender)
```

You can then append the xcm from the payload to this base xcm. If the xcm payload fails to be parsed, append nothing and send to asset hub anyway, this will cause the assets to get trapped on asset hub.